src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs

```diff
@@ -376,6 +376,146 @@ public static VersionRange Combine(IEnumerable<VersionRange> ranges, IVersionCom
         }
 
         /// <summary>
+        /// Returns the greatest common range that satisfies all given ranges.
+        /// </summary>
+        public static VersionRange CommonSubSet(IEnumerable<VersionRange> ranges)
+        {
+            return CommonSubSet(ranges, VersionComparer.Default);
+        }
+
+        /// <summary>
+        /// Returns the greatest common range that satisfies all given ranges.
+        /// </summary>
+        public static VersionRange CommonSubSet(IEnumerable<VersionRange> ranges, IVersionComparer comparer)
```

Comment:
I think this entire method could be simplified, finding the innermost set is easier than finding the outermost set (combine). In combine this pattern is used to verify there are no gaps.

You should able to: 

1. Find the highest lower bound
1. Find the lowest higher bound
1. Set the bound to exclusive if any matches on that version have it set to exclusive
1. Check if the range is valid

Here if there are any gaps, for example  [1.0.0, 2.0.0]  [4.0.0, 5.0.0]  the result will be [4.0.0, 2.0.0] which is invalid.